src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java

```diff
@@ -0,0 +1,184 @@
+
+    private static final String ERROR_AUTHENTICATING_GITHUB_APP = "Couldn't authenticate with GitHub app ID %s";
+
+    @NonNull
+    private final String appID;
+
+    @NonNull
+    private final Secret privateKey;
+
+    private String apiUri;
+
+    @DataBoundConstructor
+    @SuppressWarnings("unused") // by stapler
+    public GitHubAppCredentials(
+            CredentialsScope scope,
+            String id,
+            @CheckForNull String description,
+            @NonNull String appID,
+            @NonNull Secret privateKey
+    ) {
+        super(scope, id, description);
+        this.appID = appID;
+        this.privateKey = privateKey;
+    }
+
+    public String getApiUri() {
+        return apiUri;
+    }
+
+    @DataBoundSetter
+    public void setApiUri(String apiUri) {
+        this.apiUri = apiUri;
+    }
+
+    @NonNull
+    public String getAppID() {
+        return appID;
+    }
+
+    @NonNull
+    public Secret getPrivateKey() {
+        return privateKey;
+    }
+
+    @SuppressWarnings("deprecation") // preview features are required for GitHub app integration, GitHub api adds deprecated to all preview methods
+    static String generateAppInstallationToken(String appId, String appPrivateKey, String apiUrl) {
+        try {
+            String jwtToken = createJWT(appId, appPrivateKey);
+            GitHub gitHubApp = new GitHubBuilder().withEndpoint(apiUrl).withJwtToken(jwtToken).build();
+
+            GHApp app = gitHubApp.getApp();
+
+            List<GHAppInstallation> appInstallations = app.listInstallations().asList();
+            if (!appInstallations.isEmpty()) {
+                GHAppInstallation appInstallation = appInstallations.get(0);
+                GHAppInstallationToken appInstallationToken = appInstallation
+                        .createToken(appInstallation.getPermissions())
+                        .create();
+
+                return appInstallationToken.getToken();
+            }
+        } catch (IOException e) {
+            throw new IllegalArgumentException(String.format(ERROR_AUTHENTICATING_GITHUB_APP, appId), e);
+        }
+        throw new IllegalArgumentException(String.format(ERROR_AUTHENTICATING_GITHUB_APP, appId));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @NonNull
+    @Override
+    public Secret getPassword() {
+        if (Util.fixEmpty(apiUri) == null) {
+            apiUri = "https://api.github.com";
+        }
+
+        String appInstallationToken = generateAppInstallationToken(appID, privateKey.getPlainText(), apiUri);
+
+        return Secret.fromString(appInstallationToken);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @NonNull
+    @Override
+    public String getUsername() {
+        return appID;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Extension
+    public static class DescriptorImpl extends BaseStandardCredentialsDescriptor {
```

Comment:
Add '@Symbol'?

Reply:
What for? I've named the credential so that naturally the calculated symbol will be `githubApp`. Does adding a symbol give another benefit?

Reply:
It would be `gitHubApp` now, no? It is not a blocker, but maybe `githubApp` would be better in this case

Reply:
@timja 
I would say it is a convention.  It makes the value explicit and easy to find.
For instance, because of `@Symbol` I can easily say that `gitHubApp` is the agreed on capitalization:

https://github.com/jenkinsci/github-branch-source-plugin/blob/b953513e8d49c1ade43bdeb472b677d9279dcbc0/src/main/java/org/jenkinsci/plugins/github_branch_source/ExcludeArchivedRepositoriesTrait.java#L38

https://github.com/jenkinsci/github-branch-source-plugin/blob/aa83a7713c28b4a0a144f84081501f4c93e4c236/src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java#L154